substratevm/src/com.oracle.svm.core.graal.llvm/src/com/oracle/svm/core/graal/llvm/SubstrateNodeLLVMBuilder.java

```diff
@@ -59,6 +85,102 @@ public void emitCGlobalDataLoadAddress(CGlobalDataLoadAddressNode node) {
     @Override
     public Variable emitReadReturnAddress() {
         LLVMValueRef returnAddress = getLIRGeneratorTool().getBuilder().buildReturnAddress(getLIRGeneratorTool().getBuilder().constantInt(0));
-        return new LLVMUtils.LLVMVariable(returnAddress);
+        return new LLVMVariable(returnAddress);
+    }
+
+    private SubstrateLLVMGenerator getGenerator() {
+        return (SubstrateLLVMGenerator) gen;
+    }
+
+    @Override
+    protected int getParamIndex(int index) {
+        int offset = (getGenerator().isEntryPoint() ? 0 : LLVMFeature.SPECIAL_REGISTER_COUNT);
+        return offset + index;
+    }
+
+    @Override
+    protected LLVMValueRef[] getCallArguments(NodeInputList<ValueNode> arguments, CallingConvention.Type callType, ResolvedJavaMethod targetMethod) {
+        LLVMValueRef[] args = super.getCallArguments(arguments, callType, targetMethod);
+        return gen.getCallArguments(args, callType, targetMethod);
+    }
+
+    protected LLVMValueRef emitCondition(LogicNode condition) {
+        if (condition instanceof SafepointCheckNode) {
+            LLVMValueRef threadData = getGenerator().getSpecialRegister(LLVMFeature.THREAD_POINTER_INDEX);
+            threadData = gen.getBuilder().buildIntToPtr(threadData, gen.getBuilder().rawPointerType());
+            LLVMValueRef safepointCounterAddr = gen.getBuilder().buildGEP(threadData, gen.getBuilder().constantInt(Math.toIntExact(Safepoint.getThreadLocalSafepointRequestedOffset())));
+            LLVMValueRef safepointCount = gen.getBuilder().buildAtomicSub(safepointCounterAddr, gen.getBuilder().constantInt(1));
+            return gen.getBuilder().buildIsNull(safepointCount);
+        }
+        return super.emitCondition(condition);
+    }
+
+    @Override
+    protected LLVMValueRef emitCall(Invoke i, LoweredCallTargetNode callTarget, LLVMValueRef callee, long patchpointId, LLVMValueRef... args) {
```

Comment:
Perhaps I'm nitpicking, but I don't like single-letter names for parameters, in particular `i` which is really common as loop index.

Reply:
True, `invoke` would be more appropriante here

Reply:
Fixed.